File: modules/student-experience/src/main/java/com/campusos/portal/payload/LeavePayload.java
package com.campusos.portal.payload;

import com.campusos.portal.domain.RequestType;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

public class LeavePayload implements RequestPayload {
    public enum LeaveType { MEDICAL, PERSONAL, EVENT }

    /* ---- frozen contract fields ---- */
    public LeaveType leaveType;
    public String from;
    public String to;
    public String reason;

    /* ---- system-derived ---- */
    public Sys sys = new Sys();

    public static class Sys {
        public int dayCount;
        public int balanceAtSubmit;
        public Double attendanceBefore;
        public Double attendanceAfter;
        public List<String> datesMutated = new ArrayList<>();
        public String validation;
    }

    @Override public RequestType type() { return RequestType.LEAVE; }

    @Override public String title() {
        return leaveType.name().charAt(0) + leaveType.name().substring(1).toLowerCase() + " leave";
    }

    @Override public String subtitle() {
        return from + " → " + to + " · " + sys.dayCount + " day(s) · " + reason;
    }

    @Override public List<Artifact> artifacts() {
        List<Artifact> out = new ArrayList<>();
        if (sys.attendanceAfter != null) {
            out.add(Artifact.of("ATTENDANCE", "Attendance",
                    sys.attendanceBefore + "% → " + sys.attendanceAfter + "%"));
            out.add(Artifact.of("DAYS", "Days marked APPROVED_LEAVE",
                    String.valueOf(sys.datesMutated.size())));
        }
        return out;
    }

    @Override public void validate() {
        if (leaveType == null) throw new IllegalArgumentException("leaveType is required");
        if (reason == null || reason.isBlank()) throw new IllegalArgumentException("reason is required");
        LocalDate a = LocalDate.parse(from), b = LocalDate.parse(to);
        if (b.isBefore(a)) throw new IllegalArgumentException("to-date is before from-date");
        if (a.plusDays(30).isBefore(b)) throw new IllegalArgumentException("leave longer than 30 days");
    }
}

File: modules/student-experience-js/src/payload/LeavePayload.js
import { RequestType, LeaveType } from '../domain/enums.js';
import { parseDate, epochDay, IllegalArgumentException } from './RequestPayload.js';
import { artifact } from './Artifact.js';

/** Typed DTO for LEAVE. Plain object shape, matching the Java class field-for-field. */
export class LeavePayload {
  constructor() {
    this.leaveType = null;
    this.from = null;
    this.to = null;
    this.reason = null;
    this.sys = { dayCount: 0, balanceAtSubmit: 0, attendanceBefore: null, attendanceAfter: null,
      datesMutated: [], validation: null };
  }

  type() { return RequestType.LEAVE; }

  title() {
    const t = this.leaveType;
    return t.charAt(0) + t.slice(1).toLowerCase() + ' leave';
  }

  subtitle() { return `${this.from} → ${this.to} · ${this.sys.dayCount} day(s) · ${this.reason}`; }

  artifacts() {
    const out = [];
    if (this.sys.attendanceAfter !== null) {
      out.push(artifact('ATTENDANCE', 'Attendance', `${this.sys.attendanceBefore}% → ${this.sys.attendanceAfter}%`));
      out.push(artifact('DAYS', 'Days marked APPROVED_LEAVE', String(this.sys.datesMutated.length)));
    }
    return out;
  }

  handledBy() { return null; }

  validate() {
    if (!this.leaveType || !Object.values(LeaveType).includes(this.leaveType)) {
      throw new IllegalArgumentException('leaveType is required');
    }
    if (!this.reason || !this.reason.trim()) throw new IllegalArgumentException('reason is required');
    const a = parseDate('from-date', this.from);
    const b = parseDate('to-date', this.to);
    if (epochDay(b) < epochDay(a)) throw new IllegalArgumentException('to-date is before from-date');
    if (epochDay(b) - epochDay(a) > 30) throw new IllegalArgumentException('leave longer than 30 days');
  }

  static fromJSON(o) { return Object.assign(new LeavePayload(), o, { sys: { ...new LeavePayload().sys, ...o.sys } }); }
}
